utils/global-state-update-gen/src/generic.rs

```diff
@@ -0,0 +1,284 @@
+        }
+    }
+}
+
+fn update_auction_state<T: StateReader>(
+    state: &mut StateTracker<T>,
+    accounts: &[AccountConfig],
+    only_listed_validators: bool,
+) {
+    // Read the old SeigniorageRecipientsSnapshot
+    let (validators_key, old_snapshot) = state.read_snapshot();
+
+    // Create a new snapshot based on the old one and the supplied validators.
+    let new_snapshot = if only_listed_validators {
+        gen_snapshot_only_listed(
+            *old_snapshot.keys().next().unwrap(),
+            old_snapshot.len() as u64,
+            accounts,
+        )
+    } else {
+        gen_snapshot_from_old(old_snapshot.clone(), accounts)
+    };
+
+    if new_snapshot != old_snapshot {
+        // Save the write to the snapshot key.
+        state.write_entry(
+            validators_key,
+            StoredValue::from(CLValue::from_t(new_snapshot.clone()).unwrap()),
+        );
+
+        let validators_diff = validators_diff(&old_snapshot, &new_snapshot);
+
+        add_and_remove_bids(
+            state,
+            &validators_diff,
+            &new_snapshot,
+            only_listed_validators,
+        );
+
+        state.remove_withdraws(&validators_diff.removed);
+    }
+}
+
+/// Generates a new `SeigniorageRecipientsSnapshot` based on:
+/// - The starting era ID (the era ID at which the snapshot should start).
+/// - Count - the number of eras to be included in the snapshot.
+/// - The list of configured accounts.
+fn gen_snapshot_only_listed(
+    starting_era_id: EraId,
+    count: u64,
+    accounts: &[AccountConfig],
+) -> SeigniorageRecipientsSnapshot {
+    let mut new_snapshot = BTreeMap::new();
+    let mut era_validators = BTreeMap::new();
+    for account in accounts {
+        // don't add validatord with zero stake to the snapshot
+        let stake = match account.stake {
+            Some(stake) if stake != U512::zero() => stake,
+            _ => continue,
+        };
+        let seigniorage_recipient =
+            SeigniorageRecipient::new(stake, Default::default(), Default::default());
+        let _ = era_validators.insert(account.public_key.clone(), seigniorage_recipient);
+    }
+    for era_id in starting_era_id.iter(count) {
+        let _ = new_snapshot.insert(era_id, era_validators.clone());
+    }
+
+    new_snapshot
+}
+
+/// Generates a new `SeigniorageRecipientsSnapshot` by modifying the stakes listed in the old
+/// snaphot according to the supplied list of configured accounts.
+fn gen_snapshot_from_old(
+    mut snapshot: SeigniorageRecipientsSnapshot,
+    accounts: &[AccountConfig],
+) -> SeigniorageRecipientsSnapshot {
+    let stakes_map: BTreeMap<_, _> = accounts
+        .iter()
+        .filter_map(|acc| acc.stake.map(|stake| (acc.public_key.clone(), stake)))
+        .collect();
+
+    for recipients in snapshot.values_mut() {
+        let mut to_drop = vec![];
+        for (public_key, recipient) in recipients.iter_mut() {
```

Comment:
`BTreeMap::retain` also allows modifying values. Might be more concise.

Reply:
Fixed in [8c25879](https://github.com/casper-network/casper-node/pull/3195/commits/8c25879b795846909ecad82d3f0d2b28ed137eb8)